docs/tealscript-migration.md

```diff
@@ -0,0 +1,587 @@
+forming the transaction. All lsigs are approved unless an error occurs. Algorand TypeScript also requires implementation of the `program` method but it may not take an arguments and must return a `boolean` or `uint64` indicating whether the transaction is approved or not.
+
+##### TEALScript
+
+```ts
+class DangerousPaymentLsig extends LogicSig {
+  logic(amt: uint64) {
+    assert(this.txn.amount === amt)
+  }
+}
+```
+
+##### Algorand TypeScript
+
+TODO: How do you get args?
+
+```ts
+class DangerousPaymentLsig extends LogicSig {
+  program() {
+    return this.txn.amount === amt
+  }
+}
+```
+
+### Template Variables
+
+In TEALScript, template variables must be properties of a contract. In Algorand TypeScript, they can be defined like any other variable.
+
+###### TEALScript
+
+```ts
+class AppCaller extends LogicSig {
+  APP_ID = TemplateVar<AppID>();
+
+  logic(): void {
+    assert(this.txn.applicationID === this.APP_ID);
+  }
+}
+```
+
+###### Algorand TypeScript
+
+```ts
+class AppCaller extends LogicSig {
+  logic(): void {
+    assert(this.txn.applicationID === TemplateVar<uint64>('APP_ID'));
+  }
+}
+```
+
+### Importing
+
+In TEALScript, all of the type are injecting into the global namespace. This means no importing is required for most functions and objects.
+Algorand Typescript, however, requires explicit importing of every type, allowing for better LSP discovery.
+
+##### TEALScript
+
+```ts
+import { LogicSig } from '@algorandfoundation/tealscript'
+
+class AppCaller extends LogicSig {
+  logic(): void {
+    // No need to import assert
+    assert(this.txn.applicationID === 1234);
+  }
+}
+```
+
+##### Algorand TypeScript
+
+```ts
+import { LogicSig, Txn, assert, uint64 } from '@algorandfoundation/tealscript'
+
+class AppCaller extends LogicSig {
+  logic(): uint64 {
+    assert(Txn.applicationID === 1234);
+
+    return 1
+  }
+}
+```
+
+### Numerical Types
+
+#### `number` Type
+
+Both TEALScript and Algorand TypeScript have a `uint64` type, but Algorand TypeScript disallows any types to be resolved as `number`. This
+means all arithmetic values must be explicitly typed as `uint64`, otherwise they will have the `number` type which is not allowed.
+
+##### TEALScript
+
+```ts
+add(a: uint64, b: uint64): uint64 {
+  // Type not needed for sum
+  const sum = a + b;
+  return sum;
+}
+```
+
+##### Algorand TypeScript
+
+```ts
+add(a: uint64, b: uint64): uint64 {
+  // The type is required for sum
+  const sum: uint64 = a + b;
+  return sum;
+}
+```
+
+#### UintN types
+
+TEALScript supports typed numeric literals for most common uint types, such as `uint8`, `uint16`, `uint256`, etc. In Algorand TypeScript,
+the UintN constructors must be used.
+
+##### TEALScript
+
+```ts
+addOne(n: uint256): uint256 {
+  const one: uint256 = 1;
+  const sum = n + one;
+  return sum;
+}
+```
+
+##### Algorand TypeScript
+
+```ts
+addOne(n: UintN256): UintN256 {
```

Comment:
We can use `.native` property of `UintN256` to get `bigint` representation. So the example could be written as 
```typescript
addOne(n: UintN256): UintN256 {
  const one = 1n
  // Need to explicitly use UintN256 constructor to get uint256 and use `.native` to perform arithmetic
  const sum = new UintN256(n.native + one)
  return sum
}
```